activesupport/lib/active_support/encryptor/jwe.rb

```diff
@@ -0,0 +1,157 @@
+          auth_data_length = [auth_data.length * 8].pack('Q>')
+          mac_input = [auth_data, iv, ciphertext, auth_data_length].join
+          mac = generate_digest(mac_input, mac_key)
+          auth_tag = mac[0...(mac.length / 2)]
+        end
+
+        auth_data << '.' << [jwe_encrypted_key, iv, ciphertext, auth_tag].map { |a| encode a }.join('.')
+      end
+
+      def decrypt(encrypted_message)
+        parts = encrypted_message.split('.')
+        auth_data = parts.shift
+        encrypted_key, iv, ciphertext, auth_tag = parts.map { |a| decode a }
+        valid_header?(@serializer.load decode(auth_data))
+        valid_key? encrypted_key
+
+        cipher = new_cipher
+        cipher.decrypt
+        if gcm?
+          cipher.key = decrypt_content_encryption_key
+          cipher.iv = iv
+          cipher.auth_tag = auth_tag
+          cipher.auth_data = auth_data
+        else
+          mac_key, cipher.key = derive_mac_and_enc_key(decrypt_content_encryption_key)
+          auth_data_length = [auth_data.length * 8].pack("Q>")
+          mac_input = [auth_data, iv, ciphertext, auth_data_length].join
+          raise InvalidMessage unless untampered?(auth_tag, mac_input, mac_key)
+          cipher.iv = iv
+        end
+
+        @serializer.load(cipher.update(ciphertext) + cipher.final)
+      rescue OpenSSLCipherError, TypeError, ArgumentError
+        raise InvalidMessage
+      end
+
+      private
+        CIPHER_ENC_PAIRS = { 'aes-256-gcm' => 'A256GCM', 'aes-128-gcm' => 'A128GCM', 'aes-256-cbc' => 'A256CBC-HS512', 'aes-128-cbc' => 'A128CBC-HS256' }
+
+        def gcm?
+          ['aes-128-gcm', 'aes-256-gcm'].include?(@cipher)
+        end
+
+        def valid_header?(head)
+          raise InvalidMessage unless head.stringify_keys == header
+        end
+
+        def valid_key?(key)
+          raise InvalidMessage unless key.empty?
+        end
+
+        def new_cipher
+          OpenSSL::Cipher::Cipher.new @cipher
+        end
+
+        def encryption_algorithm
+          (@options[:enc] || @options[:cipher]).tap do |enc_or_cipher|
+            if enc_or_cipher.nil?
+              return 'aes-256-cbc'
+            elsif CIPHER_ENC_PAIRS.keys.include?(enc_or_cipher.to_s)
+              return enc_or_cipher.to_s
+            elsif CIPHER_ENC_PAIRS.values.include?(enc_or_cipher.to_s)
+              return CIPHER_ENC_PAIRS.key(enc_or_cipher.to_s)
+            else
+              raise UnexpectedAlgorithm, 'Unknown Encryption Algorithm.'
+            end
+          end
+        end
+
+        def key_encryption_algorithm
+          if @options[:alg].nil? || @options[:alg] == 'dir'
+            'dir'
+          else
+            raise UnexpectedAlgorithm, 'Unknown Key Encryption Algorithm.'
+          end
+        end
+
+        def jwe_encrypted_key
+          ''
+        end
+
+        def content_encryption_key
+          @secret
+        end
+
+        def decrypt_content_encryption_key
+          @secret
+        end
```

Comment:
Let’s not predict the future and add only what we need now.

> Den 24/08/2015 kl. 22.34 skrev Siddharth Bhatore <notifications@github.com>:
> 
> In activesupport/lib/active_support/encryptor/jwe.rb <https://github.com/sbhatore/rails/pull/4#discussion_r37800849>:
> 
> > +          else
> > +            raise UnexpectedAlgorithm, 'Unknown Key Encryption Algorithm.'
> > +          end
> > +        end
> > +
> > +        def jwe_encrypted_key
> > +          ''
> > +        end
> > +
> > +        def content_encryption_key
> > +          @secret
> > +        end
> > +
> > +        def decrypt_content_encryption_key
> > +          @secret
> > +        end
> I added these methods so that it becomes easier to add key encryption algorithms in future. For now, they seem unnecessary as I am using direct encryption. Also, the name of the functions are standard JWE names. JWE encryption key, Content Encryption Key, etc. will be very useful once I implement other key encryption algorithms. However, they can be removed for now if it seems necessary.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/sbhatore/rails/pull/4/files#r37800849>.
> 

--
Kasper